packages/core/database/src/migrations/internal-migrations/document-id.ts

```diff
@@ -0,0 +1,170 @@
+    );
+
+    return res.rows;
+  },
+  async mysql(knex: Knex, params: Params) {
+    const [res] = await knex.raw(
+      `
+    SELECT :tableName:.id as id, group_concat(DISTINCT :inverseJoinColumn:) as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :tableName:.id, :joinColumn:
+    LIMIT 1;
+  `,
+      params
+    );
+
+    return res;
+  },
+  async sqlite(knex: Knex, params: Params) {
+    return knex.raw(
+      `
+    SELECT :tableName:.id as id, group_concat(DISTINCT :inverseJoinColumn:) as other_ids
+    FROM :tableName:
+    LEFT JOIN :joinTableName: ON :tableName:.id = :joinTableName:.:joinColumn:
+    WHERE document_id IS NULL
+    GROUP BY :joinColumn:
+    LIMIT 1;
+    `,
+      params
+    );
+  },
+};
+
+const getNextIdsToCreateDocumentId = async (
+  db: Database,
+  knex: Knex,
+  {
+    joinColumn,
+    inverseJoinColumn,
+    tableName,
+    joinTableName,
+  }: {
+    joinColumn: string;
+    inverseJoinColumn: string;
+    tableName: string;
+    joinTableName: string;
+  }
+): Promise<number[]> => {
+  const res = await QUERIES[db.dialect.client as keyof typeof QUERIES](knex, {
+    joinColumn,
+    inverseJoinColumn,
+    tableName,
+    joinTableName,
+  });
+
+  if (res.length > 0) {
+    const row = res[0];
+    const otherIds = row.other_ids
+      ? row.other_ids.split(',').map((v: string) => parseInt(v, 10))
+      : [];
+
+    return [row.id, ...otherIds];
+  }
+
+  return [];
+};
+
+// Migrate document ids for tables that have localizations
+const migrateDocumentIdsWithLocalizations = async (db: Database, knex: Knex, meta: Meta) => {
+  const joinColumn = `${meta.singularName.toLowerCase()}_id`;
+  const inverseJoinColumn = `inv_${meta.singularName.toLowerCase()}_id`;
+
+  let ids: number[];
+
+  do {
+    ids = await getNextIdsToCreateDocumentId(db, knex, {
+      joinColumn,
+      inverseJoinColumn,
+      tableName: meta.tableName,
+      joinTableName: `${meta.tableName}_localizations_links`,
+    });
+
+    if (ids.length > 0) {
+      await knex(meta.tableName).update({ document_id: createId() }).whereIn('id', ids);
+    }
+  } while (ids.length > 0);
+};
+
+// Migrate document ids for tables that don't have localizations
+const migrationDocumentIds = async (db: Database, knex: Knex, meta: Meta) => {
+  let res;
+  do {
+    res = await knex.select('id').from(meta.tableName).whereNull('document_id').limit(1).first();
+
+    if (res) {
+      await knex(meta.tableName).update({ document_id: createId() }).whereIn('id', res.id);
+    }
+  } while (res !== null);
+};
+
+const createDocumentIdColumn = async (knex: Knex, tableName: string) => {
+  await knex.schema.alterTable(tableName, (table) => {
+    table.string('document_id');
+  });
+};
+
+const hasLocalizationsJoinTable = async (knex: Knex, tableName: string) => {
+  return knex.schema.hasTable(`${tableName}_localizations_links`);
+};
```

Comment:
I wonder how @innerdvations, about shortening table names, is going to affect this bit. 

After the first migration to v5, I imagine the table names will have not been shortened in knex schema right?

Reply:
I will have to update it later, but we can merge this first and I'll take it into account for my feature and fix it.

I actually suspect we will find a lot of little places throughout the codebase with hardcoded names that I'm not currently aware of and can't easily discover until I start testing the shortening and see what breaks.

Reply:
indeed, I wanted to take a look when reviewing your PR!